### PR TITLE
Add brute-force attack blocking system and DOS blocking system integration tests

### LINKED ISSUE #814: Add new API security configuration tests
## Security tests

Hi team,

Recently, we have added two new blocking systems for API requests https://github.com/wazuh/wazuh/issues/5318. The first of them is the brute-force detector, this new system detects when a brute-force attack is being made against the Wazuh login. To mitigate the attack, the system will block all requests coming from that IP for a time set by the Wazuh administrator.

The second system is a DOS blocking system. This system detects when an excessive number of calls per minute are being made to the Wazuh API. This number can be set by the user.

### New security parameters

```
block_time: Blocking time for IPs detected as brute-force attack.
max_login_attempts: Maximum number of session attempts (successes and failures) in the time set in block_time parameter.
max_request_per_minute: Maximum number of request per minute that API can handle. Avoid DOS attacks.
```

### ToDo
- [x] Make tests that check the correct functionality of the brute-force attack blocking system. Provoke a block, make a request before the blocking time finishes and one after the blocking time.
- [x] Make tests that check the correct functionality of the DOS blocking system. Provoke a block, make a request within the same minute, make a request after the minute

### PR DESCRIPTION
**Pull request by mcarmona99**
**Related issue**
#814 

Hello team,

This PR closes #814 . In this PR we have tested the correct functionality of the brute-force attack blocking system and the correct functionality of the DOS blocking system. 
Brute-force b.s. test: Provoke a block, make a request before the blocking time finishes and one after the blocking time.
DOS b.s. test: Provoke a block, make a request within the same minute, make a request after the minute

**Test results**

```
/var/ossec/framework/python/bin/python3 -m pytest wazuh-qa/
tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py 
          2 passed, 2 skipped, 25 warnings

/var/ossec/framework/python/bin/python3 -m pytest wazuh-qa/tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py
          2 passed, 2 skipped, 72 warnings
```

### CODE REVIEW

File: tests/integration/test_api/test_config/test_DOS_blocking_system/data/conf.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+---
+- tags:
+  - config1
+  conf:
+    max_request_per_minute: 50
+- tags:
+  - config2
+  conf:
+    max_request_per_minute: 10
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_DOS_blocking_system/data/conf.yaml:9]: Missing new line

File: tests/integration/test_api/test_config/test_bruteforce_blocking_system/data/conf.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    port: 55000
+    block_time: 10
+    max_login_attempts: 10
+- tags:
+  - config2
+  conf:
+    host: 0.0.0.0
+    port: 55000
+    block_time: 6
+    max_login_attempts: 3
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_bruteforce_blocking_system/data/conf.yaml:15]: Missing new line

File: tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    Provoke a block, make a request within the same minute, make a request after the minute.
+
+    Parameters
+    ----------
+    tags_to_apply : set
+        Run test if match with a configuration identifier, skip otherwise.
+    """
+    check_apply_test(tags_to_apply, get_configuration['tags'])
+    max_request_per_minute = get_configuration['conf']['max_request_per_minute']
+
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py:52]: Too many blank lines

File: tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    ----------
+    tags_to_apply : set
+        Run test if match with a configuration identifier, skip otherwise.
+    """
+    check_apply_test(tags_to_apply, get_configuration['tags'])
+    host = get_configuration['conf']['host']
+    port = get_configuration['conf']['port']
+    block_time = get_configuration['conf']['block_time']
+    max_login_attempts = get_configuration['conf']['max_login_attempts']
+
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py:56]: Too many blank lines

File: tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    check_apply_test(tags_to_apply, get_configuration['tags'])
+    host = get_configuration['conf']['host']
+    port = get_configuration['conf']['port']
+    block_time = get_configuration['conf']['block_time']
+    max_login_attempts = get_configuration['conf']['max_login_attempts']
+    
+
+    # PUT configuration for security.yaml
+    api_details = get_api_details()
+    data = {'block_time': block_time}
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py:60]: data = {'block_time': block_time, 'max_login_attempts': max_login_attempts, 'max_request_per_minute': 300}

File: tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    data['max_login_attempts'] = max_login_attempts
+    data['max_request_per_minute'] = 300   # for a different test
+    put_response = requests.put(api_details['base_url'] + '/security/config', json=data, headers=api_details['auth_headers'], verify=False)
+    assert put_response.status_code == 200, f'Expected status code was 200, ' \
+        f'but {put_response.status_code} was returned. \nFull response: {put_response.text}'
+
+    # Provoke a block from an unknown IP (default: 5 tries => ip blocked)
+    api_details = get_api_details(host=host, port=port)
+    api_details['base_url'] += '/security/user/authenticate'
+    for i in range(max_login_attempts-1):
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py:70]: Unused variable i, replace with _

File: tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        get_response = requests.get(api_details['base_url'], headers=api_details['auth_headers'], verify=False)
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py:71]: Remove unused variable get_response. requests.get(api_details['base_url'], headers=api_details['auth_headers'], verify=False)

File: tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+    # Request before blocking time expires. (5th try)
+    get_response = requests.get(api_details['base_url'], headers=api_details['auth_headers'], verify=False)
+    assert get_response.status_code == 400, f'Expected status code was 400, ' \
+        f'but {get_response.status_code} was returned. \nFull response: {get_response.text}'
+
+    # Request after time expires.
+    time.sleep(block_time) # 300 = default blocking time
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py:79]: At least two spaces before inline comment

File: tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    block_time = get_configuration['conf']['block_time']
+    max_login_attempts = get_configuration['conf']['max_login_attempts']
+    
+
+    # PUT configuration for security.yaml
+    api_details = get_api_details()
+    data = {'block_time': block_time}
+    data['max_login_attempts'] = max_login_attempts
+    data['max_request_per_minute'] = 300   # for a different test
+    put_response = requests.put(api_details['base_url'] + '/security/config', json=data, headers=api_details['auth_headers'], verify=False)
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py:63]: Line too long > 120 characters

File: tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    time.sleep(block_time) # 300 = default blocking time
+    get_response = requests.get(api_details['base_url'], headers=api_details['auth_headers'], verify=False)
+
+    # After blocking time, status code will be 401 again (unauthorized)
+    assert get_response.status_code == 401, f'Expected status code was 401, ' \
+        f'but {get_response.status_code} was returned. \nFull response: {get_response.text}'
+
+    # DELETE configuration for security.yaml
+    api_details = get_api_details()
+    delete_response = requests.delete(api_details['base_url'] + '/security/config', json=data, headers=api_details['auth_headers'], verify=False)
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py:88]: Line too long > 120 characters

File: tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+
+# Tests
+
+@pytest.mark.parametrize('tags_to_apply', [
+    {'config1'},
+    {'config2'}
+])
+def test_bruteforce_blocking_system(tags_to_apply, get_configuration, configure_api_environment, restart_api,
+                               wait_for_start, get_api_details):
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py:40]: Continuation line under-indented

File: tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        Run test if match with a configuration identifier, skip otherwise.
+    """
+    check_apply_test(tags_to_apply, get_configuration['tags'])
+    max_request_per_minute = get_configuration['conf']['max_request_per_minute']
+    
+
+    # PUT configuration for security.yaml
+    api_details = get_api_details()
+    data = {'max_request_per_minute': max_request_per_minute}
+    put_response = requests.put(api_details['base_url'] + '/security/config', json=data, headers=api_details['auth_headers'], verify=False)
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py:57]: Line too long > 120 characters

File: tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    time.sleep(60) # 60 = 1 minute
+    get_response = requests.get(api_details['base_url'], headers=api_details['auth_headers'], verify=False)
+
+    # After blocking time, status code will be 200 again 
+    assert get_response.status_code == 200, f'Expected status code was 200, ' \
+        f'but {get_response.status_code} was returned. \nFull response: {get_response.text}'
+
+    # DELETE configuration for security.yaml
+    api_details = get_api_details()
+    delete_response = requests.delete(api_details['base_url'] + '/security/config', json=data, headers=api_details['auth_headers'], verify=False)
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py:81]: Line too long > 120 characters

File: tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    for i in range(max_request_per_minute):
+        get_response = requests.get(api_details['base_url'], headers=api_details['auth_headers'], verify=False)
+
+    # Request within the same minute
+    get_response = requests.get(api_details['base_url'], headers=api_details['auth_headers'], verify=False)
+    assert get_response.status_code == 429, f'Expected status code was 429, ' \
+        f'but {get_response.status_code} was returned. \nFull response: {get_response.text}'
+
+    # Request after the minute.
+    time.sleep(60) # 60 = 1 minute
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py:72]: At least two spaces before inline comment

File: tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    return request.param
+
+
+# Tests
+
+@pytest.mark.parametrize('tags_to_apply', [
+    {'config1'},
+    {'config2'}
+])
+def test_DOS_blocking_system(tags_to_apply, get_configuration, configure_api_environment, restart_api,
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py:39]: Continuation line under-indented

File: tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    api_details = get_api_details()
+    data = {'max_request_per_minute': max_request_per_minute}
+    put_response = requests.put(api_details['base_url'] + '/security/config', json=data, headers=api_details['auth_headers'], verify=False)
+    assert put_response.status_code == 200, f'Expected status code was 200, ' \
+        f'but {put_response.status_code} was returned. \nFull response: {put_response.text}'
+
+    # Provoke an api block (default: 300 requests)
+    api_details['base_url'] += '/agents'
+    for i in range(max_request_per_minute):
+        get_response = requests.get(api_details['base_url'], headers=api_details['auth_headers'], verify=False)
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py:64]: Remove unused variable get_response. requests.get(api_details['base_url'], headers=api_details['auth_headers'], verify=False)

File: tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    # PUT configuration for security.yaml
+    api_details = get_api_details()
+    data = {'max_request_per_minute': max_request_per_minute}
+    put_response = requests.put(api_details['base_url'] + '/security/config', json=data, headers=api_details['auth_headers'], verify=False)
+    assert put_response.status_code == 200, f'Expected status code was 200, ' \
+        f'but {put_response.status_code} was returned. \nFull response: {put_response.text}'
+
+    # Provoke an api block (default: 300 requests)
+    api_details['base_url'] += '/agents'
+    for i in range(max_request_per_minute):
```
Comment (AdriiiPRodri) [inline tests/integration/test_api/test_config/test_DOS_blocking_system/test_DOS_blocking_system.py:63]: Unused variable i, replace with _

**Comment by mcarmona99** [comment]
Changes on the tests due to the changes on the configuration files:
Now the configuration of the brute force attack and DOS attack blocking systems appears in api.yaml. They appeared in security.yaml before. 
Changes on the tests were needed to make them work again. In both tests, we had to put the configuration in the endpoint  /manager/api/config instead of /security/config.

File: tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+
+    # Provoke a block from an unknown IP (default: 5 tries => ip blocked)
+    api_details = get_api_details(host=host, port=port)
+    api_details['base_url'] += '/security/user/authenticate'
+    for _ in range(max_login_attempts - 2):
+        requests.get(api_details['base_url'], headers=api_details['auth_headers'], verify=False)
+
+    # Request before blocking time expires. (5th try)
+    get_response = requests.get(api_details['base_url'], headers=api_details['auth_headers'], verify=False)
+    assert get_response.status_code == 400, f'Expected status code was 400, ' \
```
Comment (davidjiglesias) [inline tests/integration/test_api/test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py:73]:
```suggestion
    assert get_response.status_code == 403, f'Expected status code was 403, ' \
```

**Pull request merged**